### PR TITLE
feat: introduce abortScaleDownDelaySeconds to control scale down of preview/canary upon abort

### PR DESCRIPTION
**Pull request by huikang**
Address #941 #793

### CODE REVIEW

**Comment by huikang** [comment]
Hi, @jessesuen , could you please review this PR at your convenience? Thanks.

**Comment by jessesuen** [comment]
Sure I'll take a look after v1.0 release this week. This will need to wait for v1.1 since v1.0 is about to release.

**Comment by huikang** [comment]
Hi, @jessesuen , I added two test cases in e2e. Please take a look when you can. Thanks.

**Comment by jessesuen** [comment]
Jad a chance to look at this. The use case is valid, but I'd like to tweak the spec a bit to make this fit with the current design. Some considerations:

1. If we scale down the canary/desired ReplicaSet immediately upon the abort, there is a possibility of 500 routing errors due to the fact that mesh/ingress controllers can take time to shift traffic. Usually, this happens within seconds, but in general, you want to give it 30s minimum.

2. I think the time frame in which to scale down the canary after abort should be configurable. A simple boolean would make retrying an aborted rollout go through the whole scale up process and pod provisioning phase.

Therefore, I am suggesting the following change:

Instead of a `scaleDownOnAbort: true` boolean. I think this should be a duration field, which acts similarly to `spec.strategy.canary.scaleDownDelaySeconds`.

```yaml
spec:
  strategy:
    canary:
      abortScaleDownDelay: 1h
``` 

When a rollout aborts, it would immediately attach a deadline timestamp in the future to the ReplicaSet. At that timestamp, the rollout controller will scaledown the canary stack. During that delay, traffic would have already shifted back to the stable stack.

Having a delay also allows Rollouts to be retried much easier/quicker. If the Rollout is retried within the abortScaleDownDelay, then there is no time wasted provisioning pods because the ReplicaSet of the canary is already running.

**Comment by huikang** [comment]
@jessesuen , thanks for detailing the proposed design. I agree that having a scaledown delay duration would improve the traffic shitting cases. Following your example of canary, I think the spec for b/g would be

```
spec:
  strategy:
    blueGreen:
      abortScaleDownDelay: 30s
```

Is my understanding correct?

**Comment by huikang** [comment]
Hi, @jessesuen , please take a look at the PR at your convenience. Thanks.

File: rollout/replicaset.go [hunk 1]
```diff
@@ -120,10 +119,53 @@ func (c *rolloutContext) reconcileNewReplicaSet() (bool, error) {
 	if err != nil {
 		return false, err
 	}
+
+	abortScaleDownDelaySeconds := time.Duration(defaults.GetAbortScaleDownDelaySecondsOrDefault(c.rollout))
+	if c.isScaleDownOnabort() {
+		c.log.Info("scale down on abort for the newRS")
```
Comment (ngms06) [inline rollout/replicaset.go:125]: Would it be helpful to add the newRS being scaled down to the log line?

File: rollout/replicaset.go [hunk 2]
```diff
@@ -120,10 +119,53 @@ func (c *rolloutContext) reconcileNewReplicaSet() (bool, error) {
...
+
+		// if the newRS has scale down annotation and to be scaled down now, set
```
Comment (ngms06) [inline rollout/replicaset.go:127]: Is the comma supposed to be after set and not now at the end of the comment?

File: rollout/replicaset.go [hunk 3]
```diff
@@ -120,10 +119,53 @@ func (c *rolloutContext) reconcileNewReplicaSet() (bool, error) {
...
+		// Update newReplicasCount = int32(0)
+		if scaleDownAtStr, ok := c.newRS.Annotations[v1alpha1.DefaultReplicaSetScaleDownDeadlineAnnotationKey]; ok {
+			c.log.Infof("New rs '%s' has scaledown delay annotation: %s", c.newRS.Name, scaleDownAtStr)
```
Comment (ngms06) [inline rollout/replicaset.go:130]: The comment says "rs has **delay** annotation, but the previous line checks for **deadline** annotation (DefaultReplicaSetScaleDownDeadlineAnnotationKey). Is that correct?

File: rollout/replicaset.go [hunk 4]
```diff
@@ -120,10 +119,53 @@ func (c *rolloutContext) reconcileNewReplicaSet() (bool, error) {
...
+			scaleDownAtTime, err := time.Parse(time.RFC3339, scaleDownAtStr)
+			if err != nil {
+				c.log.Warnf("Unable to read scaleDownAt label on rs '%s'", c.newRS.Name)
+			} else {
+				now := metav1.Now()
+				scaleDownAt := metav1.NewTime(scaleDownAtTime)
+				if scaleDownAt.After(now.Time) {
+					c.log.Infof("RS '%s' has not reached the scaleDownTime", c.newRS.Name)
```
Comment (ngms06) [inline rollout/replicaset.go:138]: Should we add scaleDownAt to log line?

File: docs/features/specification.md [hunk 5]
```diff
@@ -115,6 +115,10 @@ spec:
       # down. Defaults to nil
       scaleDownDelayRevisionLimit: 2
 
+      # Adds a delay before scaling down the p preview replicaset if update is
```
[the remark this answers is not in the document]
Comment (huikang) [reply docs/features/specification.md]: Fixed

File: rollout/replicaset.go (hunk 4 above)
Comment (huikang) [reply rollout/replicaset.go:137]: Since line 129 logs the scale down time `scaleDownAtStr`, I think we don't replicate it here.

**Review by huikang — COMMENTED** [review]
@ngms06, thanks for reviewing. The PR is updated.

File: rollout/replicaset.go (hunk 3 above)
Comment (huikang) [reply rollout/replicaset.go]: The `DefaultReplicaSetScaleDownDeadlineAnnotationKey` is correct. Updated the log message to make it clear.

File: rollout/replicaset.go (hunk 2 above)
Comment (huikang) [reply rollout/replicaset.go]: Updated the log message. Thanks.

File: rollout/replicaset.go (hunk 1 above)
Comment (huikang) [reply rollout/replicaset.go]: log message is updated. Thanks.

**Comment by jessesuen** [comment]
@huikang I'm finally reviewing this now but during my review, I realized some of my assumptions of v1.0 behavior were not correct. It appears we **_already_** scale down canary upon abort, but a bug is preventing this from happening when used with `setCanaryScale`.

So this PR will be our opportunity to get some consistency of our behavior between all the strategies (blue-green, canary, canary w/ traffic routing, etc...).

Let me test some behavior with the various options and document what the desired behavior should/will be with all the options:
* abortScaleDownDelaySeconds = nil
* abortScaleDownDelaySeconds = 0
* abortScaleDownDelaySeconds > 0

File: docs/features/specification.md [hunk 6]
```diff
@@ -295,6 +299,11 @@ spec:
           rootService: root-svc # optional
           trafficSplitName: rollout-example-traffic-split # optional
 
+      # Add a delay before scaling down the canary pods when update
+      # is aborted for canary strategy using replicas of setCanaryScale.
+      # Default is 0, meaning canary pods are not scaled down.
+      AbortScaleDownDelaySeconds: 30
```
Comment (jessesuen) [inline docs/features/specification.md:305]: abortScaleDownDelaySeconds  should be lowercased

File: pkg/apis/rollouts/v1alpha1/types.go [hunk 7]
```diff
@@ -282,6 +286,11 @@ type CanaryStrategy struct {
 	// ScaleDownDelayRevisionLimit limits the number of old RS that can run at one time before getting scaled down
 	// +optional
 	ScaleDownDelayRevisionLimit *int32 `json:"scaleDownDelayRevisionLimit,omitempty" protobuf:"varint,12,opt,name=scaleDownDelayRevisionLimit"`
+	// AbortScaleDownDelaySeconds adds a delay before scaling down the canary pods when update
+	// is aborted for canary strategy using replicas of setCanaryScale.
+	// Default is 0, meaning canary pods are not scaled down.
```
Comment (jessesuen) [inline pkg/apis/rollouts/v1alpha1/types.go:291]: Let's add a comment stating that this field is only used with traffic routing and not applicable for basic canary.

File: rollout/replicaset.go [hunk 8]
```diff
@@ -120,10 +119,52 @@ func (c *rolloutContext) reconcileNewReplicaSet() (bool, error) {
 	if err != nil {
 		return false, err
 	}
+
+	abortScaleDownDelaySeconds := time.Duration(defaults.GetAbortScaleDownDelaySecondsOrDefault(c.rollout))
```
Comment (jessesuen) [inline rollout/replicaset.go:123]: Let's move this inside the else clause since it's only used there.

**Review by jessesuen — CHANGES_REQUESTED** [review]
Sorry for the delay in reviewing this. So here is the matrix of behavior for when abort happens, and what I feel the v1.1 behavior should be:

| strategy | v1.0 behavior | abortScaleDownDelaySeconds | desired v1.1 behavior |
|---------|---------|---------|---------|
| blue-green | does not scale down | nil | scales down after 30 seconds |
| blue-green | does not scale down | 0 | does not scale down |
| blue-green | does not scale down | N | scales down after N seconds |
| basic canary | scales down |  N/A | scales down |
| canary w/ traffic routing | scales down immediately |  nil | scales down after 30 seconds |
| canary w/ traffic routing | scales down immediately |  0 | does not scale down |
| canary w/ traffic routing | scales down immediately |  N | scales down after N seconds |
| canary w/ traffic routing + setCanaryScale | does not scale down (bug) | * | should behave like above |

@huikang  - I think v1.1 should make the behavior consistent across all strategies. Basically, upon an abort, we should scale down the new replicaset for all strategies. Users can then choose to leave the new replicaset scaled up indefinitely by setting  `abortScaleDownDelaySeconds` to 0, or adjust the value to something larger (or smaller).

**Comment by huikang** [comment]
> Sorry for the delay in reviewing this. So here is the matrix of behavior for when abort happens, and what I feel the v1.1 behavior should be:
> strategy 	v1.0 behavior 	abortScaleDownDelaySeconds 	desired v1.1 behavior
> blue-green 	does not scale down 	nil 	scales down after 30 seconds
> blue-green 	does not scale down 	0 	does not scale down
> blue-green 	does not scale down 	N 	scales down after N seconds
> basic canary 	rolling update back to stable 	N/A 	rolling update back to stable
> canary w/ traffic routing 	scales down immediately 	nil 	scales down after 30 seconds
> canary w/ traffic routing 	scales down immediately 	0 	does not scale down
> canary w/ traffic routing 	scales down immediately 	N 	scales down after N seconds
> canary w/ traffic routing + setCanaryScale 	does not scale down (bug) 	* 	should behave like above
> 
> @huikang - I think v1.1 should make the behavior consistent across all strategies. Basically, upon an abort, we should scale down the new replicaset for all strategies. Users can then choose to leave the new replicaset scaled up indefinitely by setting `abortScaleDownDelaySeconds` to 0, or adjust the value to something larger (or smaller).


@jessesuen , thanks for summarizing. I will update the PR to reflect the descried behavior as listed in the table. In addition, I will update the doc to include the table.

**Comment by huikang** [comment]
Hi, @jessesuen , the PR has been updated based upon the desired behavior of v1.1. Please help review at your convenience. Thanks.

**Review by jessesuen — APPROVED** [review]
Great work!

**Comment by dudadornelles** [comment]
Was this released? Really looking forward to use this feature, thanks for the hard work folks!

**Comment by alexandresavicki** [comment]
> Was this released? Really looking forward to use this feature, thanks for the hard work folks!

I'm try to use this parameter using version 1.0.4 but i'm getting error:
```
# rollouts.argoproj.io "dummy-app" was not valid:
# * : Invalid value: "The edited file failed validation": ValidationError(Rollout.spec.strategy.canary): unknown field "abortScaleDownDelaySeconds" in io.argoproj.v1alpha1.Rollout.spec.strategy.canary
```

**Comment by jessesuen** [comment]
Not yet. It's a v1.1 feature. You can try the feature out by installing the manifests from master, which use the `:latest` images

**Comment by alexandresavicki** [comment]
@jessesuen Thanks a lot for your reply. Just one more doubt i'm looking at this parameter trying bypass some downtime when canary fails for any reason (eg.: Analysis step not match specified conditions) so canary will be aborted and rollback will be started. I understand the rollback process it's very aggressive such as traffic shift it's slow and can cause downtime in my application.
There is an another way to  bypass this scenario?

**Pull request merged**